samza-test/src/test/java/org/apache/samza/test/operator/RepartitionWindowApp.java

```diff
@@ -42,16 +43,20 @@
 
   @Override
   public void init(StreamGraph graph, Config config) {
+    graph.setDefaultKeySerde(new StringSerde());
+    // will fail with class cast in partitionBy if (optional) PageView.class param not provided
+    graph.setDefaultMsgSerde(new JsonSerde<>(PageView.class));
```

Comment:
Given the likelihood that messages are going to change in some way in the pipeline and that this is not type safe, have we considered defaulting not for the whole pipeline but by type?

One other consideration is that neglecting to set this and an explicit serializer later would give an info entry in the log and a class cast exception. It seems that it would be better to require the user to explicitly use a pass-through SerDe and error out early if not SerDe is set. Also, if we really do expose the ClassCastException to the user it would be good to augment that with some info about what to look for (e.g. make sure serde matches type).

Reply:
The major use cases for these default serdes in my mind were:
1. For the keySerde, when the key type is always the same (e.g. String/Integer).
2. For the msgSerde, when the data is avro/json/pegasus/protobuf and the Serde can figure out the schema from the message contents.
For DAGs of low to medium complexity, I don't think default serdes per type would be much more concise than specifying them inline. It's also non-trivial to infer the concrete type M for a given MessageStream at compile/planning time, so I don't know of a simple implementation other than checking the message type at runtime. Maybe we can add them later if we find that we need them?

Re: default serde == NoOpSerde vs null (and throwing on use). This version optimizes for users of Systems that do the serialization themselves. E.g., at LinkedIn, this is the case for the Kafka SystemProducer/SystemConsumer where avro is enforced and user-provided serdes are disallowed/ignored. I didn't want to make this yet another configuration. Maybe we can make this null by default and patch it internally to be NoOpSerde instead? @jmakes @nickpan47, any thoughts on this?

Agree that we should add more information around ClassCastException where possible. Currently it's hard to find the source operator for the exception from the stacktrace.

Reply:
Agree with the use cases and with these methods being optional s.t. if they add value, use them, if not, don't.

I'm not sure I understand the difference between PassthroughSerde and NoOpSerde. They sound the same to me. 

Do you think it's useful to expose a typed NoOpSerde<> so users that do their own serialization can hint the type info expected by their implementation? e.g. at LinkedIn we could enforce NoOpSerde<IndexedRecord>. I honestly didn't look close enough at the code to verify whether this is useful, just throwing it out there.

I don't think we have much data to say whether or not the defaults will be more or less useful. With avro specific records, you could imagine each external stream having a different type and therefore another serde instance. However, I think a number of jobs at LI still use generic records, and for them this might be more useful.

Reply:
@jmakes In this context, pass through == byte[] in byte[] out, whereas no-op == don't attempt to serialize at all. Maybe NoSerde is better?

The NoOpSerde is already typed in this PR. However, note that if you set the default serde, you don't get any type inference when creating input/output streams. You have to provide the type of the MessageStream contents manually and correctly based on the default serde.